infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/multiuser/oauth/IdentityProviderConfigBuilderTest.java

```diff
@@ -0,0 +1,230 @@
+    when(keycloakServiceClient.getIdentityProviderToken(anyString())).thenReturn(tokenResponse);
+    when(workspaceRuntimes.getRuntimeContext(anyString()))
+        .thenReturn(Optional.<RuntimeContext>empty());
+
+    Config resultConfig = configBuilder.buildConfig(defaultConfig, A_WORKSPACE_ID);
+    assertEquals(resultConfig.getOauthToken(), ACCESS_TOKEN);
+  }
+
+  @Test
+  public void testCreateUserConfigWhenWorkspaceIdIsNull() throws Exception {
+    when(keycloakServiceClient.getIdentityProviderToken(anyString())).thenReturn(tokenResponse);
+    Config resultConfig = configBuilder.buildConfig(defaultConfig, null);
+    assertEquals(resultConfig.getOauthToken(), ACCESS_TOKEN);
+  }
+
+  @Test
+  public void testCreateUserConfig() throws Exception {
+    when(keycloakServiceClient.getIdentityProviderToken(anyString())).thenReturn(tokenResponse);
+    Config resultConfig = configBuilder.buildConfig(defaultConfig, A_WORKSPACE_ID);
+    assertEquals(resultConfig.getOauthToken(), ACCESS_TOKEN);
+  }
+
+  @Test(expectedExceptions = {InfrastructureException.class})
+  public void testThrowOnBadScope() throws Exception {
+    when(keycloakServiceClient.getIdentityProviderToken(anyString())).thenReturn(tokenResponse);
+    when(tokenResponse.getScope()).thenReturn("bad:scope");
+    Config resultConfig = configBuilder.buildConfig(defaultConfig, A_WORKSPACE_ID);
+    assertEquals(resultConfig.getOauthToken(), ACCESS_TOKEN);
+  }
+
+  public void testRethrowOnUnauthorizedException() throws Exception {
+    doThrow(new UnauthorizedException(ServiceErrorImpl.make().withMessage("Any other message")))
+        .when(keycloakServiceClient)
+        .getIdentityProviderToken(anyString());
+    try {
+      configBuilder.buildConfig(defaultConfig, A_WORKSPACE_ID);
+    } catch (InfrastructureException e) {
+      assertEquals(e.getMessage(), SHOULD_LINK_ERROR_MESSAGE, "The exception message is wrong");
+      return;
+    }
+    fail(
+        "Should have thrown an exception with the following message: " + SHOULD_LINK_ERROR_MESSAGE);
+  }
+
+  @Test(expectedExceptions = {InfrastructureException.class})
+  public void testRethrowOnBadRequestException() throws Exception {
+    doThrow(new BadRequestException(ServiceErrorImpl.make().withMessage("Any other message")))
+        .when(keycloakServiceClient)
+        .getIdentityProviderToken(anyString());
+    configBuilder.buildConfig(defaultConfig, A_WORKSPACE_ID);
+  }
+
+  public void testRethrowOnInvalidTokenBadRequestException() throws Exception {
+    doThrow(new BadRequestException(ServiceErrorImpl.make().withMessage("Invalid token.")))
+        .when(keycloakServiceClient)
+        .getIdentityProviderToken(anyString());
+    try {
+      configBuilder.buildConfig(defaultConfig, A_WORKSPACE_ID);
+    } catch (InfrastructureException e) {
+      assertEquals(e.getMessage(), SESSION_EXPIRED_MESSAGE, "The exception message is wrong");
```

Comment:
You can match exception message in test annotation with field `expectedExceptionsMessageRegExp`. This would make the code of some of these tests clearer. And it also supports wildcards.